api/wikipediaSource.ts

```diff
@@ -0,0 +1,80 @@
+
+
+/**
+ * Fetch and return the introduction as plain text of the given Wikipedia page.
+ * Uses the MediaWiki Action API.
+ * @param pageTitle the title of the Wikipedia page, must be first-capitalized
+ * words (except for name particles) separated by '_', for instance "Leonardo_da_Vinci"
+ * @throws Error if no page corresponding to the given title is found
+ */
+async function fetchIntro(pageTitle: string): Promise<string> {
+    const searchParams: Record<string, string> = {
+        action: "query",
+        titles: pageTitle,
+        format: "json",
+        prop: "extracts",
+        exintro: "true",
+        explaintext: "true",
+        origin: "*"
+    }
+
+    const queryParams: string = new URLSearchParams(searchParams).toString()
+    const url: string = BASE_URL + ENDPOINT + queryParams
+
+    return fetch(url)
+        .then(response => response.json())
+        .then(data => {
+            if ('query' in data && 'pages' in data.query) {
+                const pages = data.query.pages
+                const pageIds: string[] = Object.keys(pages)
+                if (pageIds.length === 1 && pageIds[0] !== '-1') {
+                    const pageId: number = Number.parseInt(pageIds[0])
+                    return pages[pageId].extract
+                }
+            }
+            throw new Error(`Page with title ${pageTitle} was not found.`)
+        })
+}
+
+/**
+ * Fetch and return the source URL of the main picture of the given Wikipedia page.
+ * Uses the MediaWiki Action API.
+ * @param pageTitle the title of the Wikipedia page, must be first-capitalized
+ * words (except for name particles) separated by '_', for instance "Leonardo_da_Vinci"
+ * @param thumbSize the width in pixels of the wanted thumbnail
+ * @throws Error if no page corresponding to the given title is found
+ */
+async function fetchImageUrl(pageTitle: string, thumbSize: number): Promise<string> {
+    const searchParams: Record<string, string> = {
+        action: "query",
+        titles: pageTitle,
+        format: "json",
+        prop: "pageimages",
+        piprop: "thumbnail",
+        pithumbsize: thumbSize.toString(),
+        origin: "*",
+    };
+
+    const queryParams: string = new URLSearchParams(searchParams).toString()
+    const url: string = BASE_URL + ENDPOINT + queryParams
+
+   return fetch(url)
+       .then(response => response.json())
+       .then(data => {
+           if ('query' in data && 'pages' in data.query) {
+               const pages = data.query.pages
+               const pageIds: string[] = Object.keys(pages)
+               if (pageIds.length === 1 && pageIds[0] !== '-1') {
+                   const pageId: number = Number.parseInt(pageIds[0])
+                   if ('thumbnail' in pages[pageId]) {
+                       return pages[pageId].thumbnail.source
+                   }
+               }
+           }
+           throw new Error(`Image for page with title ${pageTitle} was not found.`)
```

Comment:
When you execute `throw new Error(...)`, you never catch it, so it doesn't do anything, right?
```suggestion
           throw new Error(`Image for page with title ${pageTitle} was not found.`)
    }).catch(error => {
            console.error('Error fetching Wikipedia page:', error);
    });
```

Reply:
You're right, I just commited your suggestions.